File: 08/piggy/construct_binary_tree_from_preorder_and_inorder_traversal.c
struct TreeNode* build(int* preorder, int pres, int pree, int* inorder, int ins, int ine) {
    struct TreeNode* root = malloc(sizeof(struct TreeNode));
    root->val = preorder[pres];

    if (pres == pree) {
        root->left = NULL;
        root->right = NULL;
        return root;
    }

    int ni = ins;

    for(; ni <= ine; ni++) {
        if(inorder[ni] == preorder[pres])
            break;
    }

    int pi = ni - ins + pres;

    root->left = build(preorder, pres + 1, pi, inorder, ins, ni - 1);
    root->right = build(preorder, pi + 1, pree, inorder, ni + 1, ine);

    return root;
} 

struct TreeNode* buildTree(int* preorder, int preorderSize, int* inorder, int inorderSize) {
    if (preorderSize == 0 || inorderSize == 0)
        return NULL;
        
    return build(preorder, 0, preorderSize - 1, inorder, 0, inorderSize - 1);
}

File: 08/pierpont/Construct Binary Tree from Preorder and Inorder Traversal.java
import java.util.HashMap;

public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(inorder==null || preorder==null || inorder.length==0 || preorder.length==0)  
        {  
            return null;  
        }  
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();  
        for(int i=0;i<inorder.length;i++)  
        {  
            map.put(inorder[i],i);  
        }  
    	
        return helper(inorder,preorder,0,inorder.length-1, 0, preorder.length-1,map); 
    }
    
    private TreeNode helper(int[] inorder, int[] preorder, int inL, int inR, int preL, int preR, 
    		HashMap<Integer, Integer> map)  
    {  
        if(inL>inR || preL>preR)  
        	return null;  
        TreeNode root = new TreeNode(preorder[preL]);  
        int index = map.get(root.val);  
        root.left = helper(inorder,preorder,inL,index-1,preL + 1,preL+index-inL,map);  
        root.right = helper(inorder,preorder,index+1,inR,preR-(inR-index),preR,map);  
        return root;
    	
    }
    
    public class TreeNode {
        int val;
    	TreeNode left;
    	TreeNode right;
    	TreeNode(int x) { val = x; }
    }
}
